Improve performance of resizing isbits Union Arrays

quinnj opened the pull request:
cc: @vtjnash 

Most of the code is in pretty good shape I think. Unfortunately my llvm is super rusty, so I haven't made the necessary changes in codegen.cpp for getindex/setindex!. I might have to do a hangout call w/ @vtjnash to remember how to navigate the llvm codegen stuff. Basically, we just need to adjust where the selector bytes are pulled from: they're stored after `a->maxsize` now instead of directly after the last array element.

Keno commented on src/codegen.cpp:
```diff
@@ -2518,7 +2518,7 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
                     Value *data = emit_bitcast(ctx, emit_arrayptr(ctx, ary, ary_ex), AT->getPointerTo());
-                    // isbits union selector bytes are stored directly after the last array element
+                    // isbits union selector bytes are stored after a->maxsize
                     Value *selidx = emit_arraylen_prim(ctx, ary);
```
This now needs to be maxsize

vtjnash replied:
```c++
Value *ndims = (nd == -1 ? emit_arrayndims(ctx, ary) : ConstantInt::get(T_int16, nd));
Value *is_vector = ctx.builder.CreateICmpEQ(ndims, ConstantInt::get(T_int16, 1));
Value *offset = emit_arrayoffset(ctx, ary);
Value *selidx_v = ctx.builder.CreateSub(emit_vectormaxsize(ctx, ary), ctx.builder.CreateZExt(T_size, offset));
Value *selidx_m = emit_arraylen(ctx, ary);
Value *seldx = ctx.builder.CreateSelect(is_vector, selidx_v, selidx_m);
Value *ptindex = ctx.builder.CreateInBoundsGEP(AT, data, selidx);
ptindex = emit_bitcast(ctx, ptindex, T_pint8);
ptindex = ctx.builder.CreateInBoundsGEP(T_int8, ptindex, offset);
```
```c++
static Value *emit_arrayndims(jl_codectx_t &ctx, const jl_cgval_t &ary)
{
    Value *flags = emit_arrayflags(ctx, ary);
    cast<LoadInst>(flags)->setMetadata(LLVMContext::MD_invariant_load, MDNode::get(md->getContext(), None));
    flags = ctx.builder.CreateLShr(flags, 2);
    flags = ctx.builder.CreateAnd(flags, 0x3FF); // (1<<10) - 1
    return flags;
}

static Value *emit_vectormaxsize(jl_codectx_t &ctx, const jl_cgval_t &ary)
{
    return emit_arraysize(ctx, ary, 2); // maxsize aliases ncols in memory layout for vector
}
```

quinnj commented on src/array.c:
```diff
@@ -710,58 +709,85 @@ STATIC_INLINE void jl_array_grow_at_beg(jl_array_t *a, size_t idx, size_t inc,
             // We could use memcpy if resizing allocates a new buffer,
             // hopefully it's not a particularly important optimization.
-            if (idx > 0 && newdata < data)
+            if (idx > 0 && newdata < data) {
```
I'm not sure this codepath is reachable; the conditions above seem to prevent the case where `inc` will ever be larger than `a->offset`, yet smaller than `(a->maxsize - n) / 2 - (a->maxsize - n) / 20`.

Just a note as I've been trying to ensure we have tests that hit all these various conditional branches.

quinnj commented on src/array.c:
```diff
@@ -895,6 +915,7 @@ STATIC_INLINE void jl_array_shrink(jl_array_t *a, size_t dec)
         //this is a julia-allocated buffer that needs to be marked
     }
     else if (a->flags.how == 2) {
+        //FIXME: what to do about isbits Union arrays w/ their type tag bytes?
```
I'm pretty sure this is currently an issue w/ isbits Union Arrays, since malloc'ed arrays are just straight up shrunk and we're not accounting for moving the type tag bits around. Still thinking through what we need to do here.

quinnj commented:
Ok, I think this is in pretty good shape now. I've added a bunch of tests to cover all the various codepaths of the various resizing strategies and they're all looking good, at least w/ using the internal Base methods to resize & `arrayref`/`arrayset` to test.

So remaining is to:
[ ]: figure out the necessary codegen.cpp magic; @vtjnash, @Keno, @JeffBezanson, if one of you has time to help, I'd really appreciate the help
[ ]: figure out what to do about `jl_array_shrink`, since it's currently not accounting for the special layout of isbits Union Arrays; technically a separate issue from the original performance issues, but since we're refactoring here anyway....
[ ]: have at least @vtjnash review the refactor here and make sure we're not doing anything super bad or fatal

quinnj commented:
Ok, just pushed a commit that should fix `jl_array_shrink` for isbits Union arrays, and I threw in a bonus fix that affected all malloced arrays that had `a->offset != 0`. I'm a bit surprised we never ran into this bug, but it's pretty easily reproducible like
```julia
julia> A = collect(1:1048577);

julia> pushfirst!(A, 0);

julia> deleteat!(A, 10:length(A));

julia> sizehint!(A, 9)
julia(62929,0x7fff944993c0) malloc: *** error for object 0x128f1f000: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug

signal (6): Abort trap: 6
in expression starting at no file:0
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Allocations: 4319571 (Pool: 4318703; Big: 868); GC: 9
Abort trap: 6
```
I guess people generally aren't doing such drastic resizing/sizehint!ing on large, malloced arrays.

Anyway, so all we need here is the codegen.cpp updates, which I'll continue to pester @Keno, @vtjnash, and @JeffBezanson about.

quinnj commented:
Bump for @keno or @vtjnash

quinnj commented on src/cgutils.cpp:
```diff
@@ -1743,6 +1748,15 @@ static Value *emit_arrayflags(jl_codectx_t &ctx, const jl_cgval_t &tinfo)
+static Value *emit_arrayndims(jl_codectx_t &ctx, const jl_cgval_t &ary)
+{
+    Value *flags = emit_arrayflags(ctx, ary);
+    cast<LoadInst>(flags)->setMetadata(LLVMContext::MD_invariant_load, MDNode::get(jl_LLVMContext, None));
```
I had to change to `jl_LLVMContext` because of undef `md` errors.

quinnj commented on src/cgutils.cpp:
```diff
@@ -1757,6 +1771,21 @@ static Value *emit_arrayelsize(jl_codectx_t &ctx, const jl_cgval_t &tinfo)
     return tbaa_decorate(tbaa_const, ctx.builder.CreateLoad(addr));
 }
 
+static Value *emit_arrayoffset(jl_codectx_t &ctx, const jl_cgval_t &tinfo)
```
I think this function should be right? Maybe there's something off?

quinnj commented on src/codegen.cpp:
```diff
@@ -241,6 +241,7 @@ static MDNode *tbaa_arrayptr;       // The pointer inside a jl_array_t
 static MDNode *tbaa_arraysize;      // A size in a jl_array_t
 static MDNode *tbaa_arraylen;       // The len in a jl_array_t
 static MDNode *tbaa_arrayflags;     // The flags in a jl_array_t
+static MDNode *tbaa_arrayoffset;     // The offset in a jl_array_t
```
I assume we need something like this?

quinnj commented on src/codegen.cpp:
```diff
@@ -2511,8 +2512,16 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
+                    Value *selidx_v = ctx.builder.CreateSub(emit_vectormaxsize(ctx, ary), ctx.builder.CreateZExt(offset, T_size));
+                    Value *selidx_m = emit_arraylen(ctx, ary);
+                    Value *selidx = ctx.builder.CreateSelect(is_vector, selidx_v, selidx_m);
+                    Value *elsize = emit_arrayelsize(ctx, ary);
```
I added these three lines since the type tag first element should be `(a->maxsize - a->offset) * a->elsize + a->offset`. Not sure if we need to add a `CreateZExt` for `elsize` or `offset` here though? (we have one above when we subtract maxsize and offset). I tried w/ and I got segfaults either way.

quinnj commented:
Update: I tried incorporating @vtjnash's suggestions, made some edits/updates, but still can't seem to get a basic `getindex` call to avoid segfaulting. @vtjnash, if you or @keno could take a look and make sure I didn't mess anything up in my edits?

quinnj commented:
Hmmmm, not sure why windows is the only platform having trouble w/ that `memmove` call (freebsd failure looks unrelated?). I'll try to double check things again today.

quinnj commented:
Alright, let's see how this latest commit fares; everything passes for me locally.

quinnj commented:
Alright, all green here (finally). I kind of feel like doing a rebase, just to clean it up, but if people would rather squash while everything's green, that'd be fine. Also, anyone want to take another look?

quinnj committed:
Refactor isbits Union arrays to treat the type tag bytes like a proper jl_array_t* with potential a->offset and a->maxsize values. Before, any resizing operations on isbits Union arrays (push, pushfirst, insert, append, deleteat, slice, etc.) incurred extra cost by having to constantly move the type tag bytes to be directly after the last array element. By treating them rather as jl_array_t* objects, sharing the a->offset and a->maxsize fields with the parent array, resizing operations now only require moving bytes around if the parent array's data itself must be moved. Fixes #27825 and #27809.

quinnj committed:
Fix issue in jl_array_shrink where the offset wasn't correctly multiplied by a->elsize, so the array's data pointer got messed up.

quinnj commented:
Ok, I squashed this manually locally and separated out just the `jl_array_shrink` fix as a separate commit (which affected all arrays, not just isbits unions). I also added a devdoc blurb about the isbits Union optimizations that might be useful for other devs to be aware of in the future.

StefanKarpinski commented:
Straight flush again. The CI gods truly smile upon you!

maleadt commented:
Is it possible this broke DataFrames? On #28216 + master (as of 88a8c865a2e95216406c241e715ed56e75032775):

```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-beta2.85 (2018-07-23 13:43 UTC)
 _/ |\__'_|_|_|\__'_|  |  jb/tvarspecificity/3605a8d0ca* (fork: 2 commits, 0 days)
|__/                   |  x86_64-pc-linux-gnu

(tools) pkg> st
using     Status `Project.toml`
  [336ed68f] CSV v0.2.5
  [a93c6f00] DataFrames v0.11.7
  [31c24e10] Distributions v0.16.1
  [c27321d9] Glob v1.1.1
  [eff96d63] Measurements v0.5.1
  [295af30f] Revise v0.5.4

julia> using DataFrames
[ Info: Recompiling stale cache file /home/tbesard/Julia/depot/compiled/v0.7/DataFrames/AR9o.ji for module DataFrames

julia> dfA = DataFrame(:a => Union{Missing,Int}[])
0×1 DataFrame


julia> push!(dfA, 1)
1×1 DataFrame
│ Row │ a │
├─────┼───┤
│ 1   │ 1 │

julia> dfB = DataFrame(:a => Union{Missing,Int}[])
0×1 DataFrame


julia> push!(dfB, 1)
1×1 DataFrame
│ Row │ a │
├─────┼───┤
│ 1   │ 1 │

julia> vcat(dfA, dfB)

signal (11): Segmentation fault
```

I've also seen it construct a DataFrame with only `missing` values (ie. where there really aren't any), without a segfault. Works fine on #28216 + 1d3c9c8651ee48ee7e27fe478d369b707c1efb10 (the earliest commit where DataFrames works again).

I'll bisect when I have some time.

quinnj commented:
Taking a look

quinnj commented:
kk, fix PR [here](https://github.com/JuliaLang/julia/pull/28243). There were two places in Base itself that I forgot to update w/ the new layout changes; we have tests for `copyto!` and `vcat` for isbits Union Arrays, which passed, but only because the former and new layouts are the same when an array doesn't have any internal padding (i.e. `a->offset == 0` and `a->maxsize == a->nrows * a->elsize`). The PR fixes and adds some more tests cases.

maleadt commented:
Thanks!

vtjnash commented:
Nice work here, Jacob. Thanks!
